### PR TITLE
Feat: Add Iris Server Support

### PR DESCRIPTION
**Pull request by Ns2Kracy**
I think most of the work has been done.Is there any reviewer?

### CODE REVIEW

**Comment by Ns2Kracy** [comment]
@deepmap-marcinr  if there is anyone available on the team who can review it?

**Comment by jamietanna** [comment]
Thanks for your first contribution and it's a big one 🙌  We'll try and review soon - there's a few things we'll want to get in following the v1.13.0 release for a patch bump, then can look at this one

**Comment by Ns2Kracy** [comment]
> 

Looking forward for it!

File: pkg/iris-middleware/oapi_validate.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+const (
+	IrisContextKey = "oapi-codegen/iris-context"
+	UserDataKey    = "oapi-codegen/user-data"
+)
+
+// OapiValidatorFromYamlFile creates a validator middleware from a YAML file path
+func OapiValidatorFromYamlFile(path string) (iris.Handler, error) {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("error reading %s: %s", path, err)
```
Comment (alexandear) [inline pkg/iris-middleware/oapi_validate.go:27]: Could we use native error wrapping?

```suggestion
		return nil, fmt.Errorf("error reading %s: %w", path, err)
```

File: pkg/iris-middleware/oapi_validate.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	}
+
+	swagger, err := openapi3.NewLoader().LoadFromData(data)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing %s as Swagger YAML: %s",
```
Comment (alexandear) [inline pkg/iris-middleware/oapi_validate.go:32]:
```suggestion
		return nil, fmt.Errorf("error parsing %s as Swagger YAML: %w",
```

File: pkg/iris-middleware/oapi_validate.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	if err != nil {
+		switch e := err.(type) {
+		case *routers.RouteError:
+			// We've got a bad request, the path requested doesn't match
+			// either server, or path, or something.
+			return errors.New(e.Reason)
+		default:
+			// This should never happen today, but if our upstream code changes,
+			// we don't want to crash the server, so handle the unexpected error.
+			return fmt.Errorf("error validating route: %s", err.Error())
```
Comment (alexandear) [inline pkg/iris-middleware/oapi_validate.go:99]:
```suggestion
			return fmt.Errorf("error validating route: %w", err)
```

File: pkg/iris-middleware/oapi_validate.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+			// Split up the verbose error by lines and return the first one
+			// openapi errors seem to be multi-line with a decent message on the first
+			errorLines := strings.Split(e.Error(), "\n")
+			return fmt.Errorf("error in openapi3filter.RequestError: %s", errorLines[0])
+		case *openapi3filter.SecurityRequirementsError:
+			return fmt.Errorf("error in openapi3filter.SecurityRequirementsError: %s", e.Error())
+		default:
+			// This should never happen today, but if our upstream code changes,
+			// we don't want to crash the server, so handle the unexpected error.
+			return fmt.Errorf("error validating request: %s", err)
```
Comment (alexandear) [inline pkg/iris-middleware/oapi_validate.go:140]:
```suggestion
			return fmt.Errorf("error validating request: %w", err)
```

**Review by alexandear — CHANGES_REQUESTED** [review]

**Review by alexandear — CHANGES_REQUESTED** [review]
Cool.

I left some comments.

**Comment by Ns2Kracy** [comment]
> Cool.
> 
> I left some comments.

I'll finish it

File: pkg/codegen/operations.go [hunk 5]
```diff
@@ -913,6 +911,12 @@ func GenerateTypesForOperations(t *template.Template, ops []OperationDefinition)
 	return buf.String(), nil
 }
 
+// GenerateIrisServer This function generates all the go code for the ServerInterface as well as
```
[the remark this answers is not in the document]
Comment (Ns2Kracy) [reply pkg/codegen/operations.go:914]: This is modified from echo, maybe no modification is needed?

**Review by alexandear — APPROVED** [review]

**Comment by Ns2Kracy** [comment]
I think all the work is basically done! Great!

**Comment by deepmap-marcinr** [comment]
Thank you for doing this.

I need to find a way of abstracting away all these different routers behind a stable interface for template code. It'd be nice if oapi-codegen could simply import someone else's code generator as a module.

**Review by jamietanna — CHANGES_REQUESTED** [review]
Will require a couple of tweaks - looks like we may be missing the changes added for #1168 (as they were completed after this PR was first raised) - happy to add them for you!

**Review by jamietanna — APPROVED** [review]

**Comment by Ns2Kracy** [comment]
> Will require a couple of tweaks - looks like we may be missing the changes added for #1168 (as they were completed after this PR was first raised) - happy to add them for you!

Thanks!

**Comment by lzap** [comment]
Just to give you little feedback, this feature pulled a ton of dependencies which are no longer 1.19 compatible. The only solution for now is to freeze oapi-codegen until our Go runtime provider will upgrade the stack. Would be awesome to separate generators from the runtime at some point.

Anyways, thanks for the project it works really well. We use just the plain good old Go standard library :-D

**Comment by jamietanna** [comment]
Thanks @lzap! Yes we're aware, we made some work towards this recently based on https://github.com/deepmap/oapi-codegen/issues/1142.

By November there's [a v2 planned (which is very minor breaking changes, should be zero work for almost all consumers)](https://github.com/deepmap/oapi-codegen/discussions/1309) which removes the transitive dependencies that you'll be seeing due to the current package structure, which will reduce the number + size of the dependency tree considerably.

If you check out [this comment](https://github.com/deepmap/oapi-codegen/discussions/1309#discussioncomment-7238363) you should be able to pin to [the proposed changes](https://github.com/deepmap/oapi-codegen/pull/1310) to see that you may be OK. Additionally, by moving to [v1.15.0: Remove issues with Go 1.21](https://github.com/deepmap/oapi-codegen/releases/tag/v1.15.0) you won't be required to use Go 1.21. The minimum required Go version is 1.20 as [Go 1.19 is no longer supported](https://endoflife.date/go).

**Comment by jamietanna** [comment]
> We use just the plain good old Go standard library :-D

You may also be interested in https://github.com/deepmap/oapi-codegen/issues/1068 which looks likely to land in Go 1.22

**Comment by lzap** [comment]
Thanks for the extra work, that had to be ton of code to do that. Awesome.

**Pull request merged**